cura/Settings/CuraContainerStack.py

```diff
@@ -0,0 +1,616 @@
+
+    ##  Overridden from ContainerStack
+    #
+    #   Since we have a fixed order of containers in the stack and this method would modify the container
+    #   ordering, we disallow this operation.
+    @override(ContainerStack)
+    def insertContainer(self, index: int, container: ContainerInterface) -> None:
+        raise Exceptions.InvalidOperationError("Cannot insert a container into Global stack")
+
+    ##  Overridden from ContainerStack
+    #
+    #   Since we have a fixed order of containers in the stack and this method would modify the container
+    #   ordering, we disallow this operation.
+    @override(ContainerStack)
+    def removeContainer(self, index: int) -> None:
+        raise Exceptions.InvalidOperationError("Cannot remove a container from Global stack")
+
+    ##  Overridden from ContainerStack
+    #
+    #   Replaces the container at the specified index with another container.
+    #   This version performs checks to make sure the new container has the expected metadata and type.
+    #
+    #   \throws Exception.InvalidContainerError Raised when trying to replace a container with a container that has an incorrect type.
+    @override(ContainerStack)
+    def replaceContainer(self, index: int, container: ContainerInterface, postpone_emit: bool = False) -> None:
+        expected_type = _ContainerIndexes.IndexTypeMap[index]
+        if expected_type == "definition":
+            if not isinstance(container, DefinitionContainer):
+                raise Exceptions.InvalidContainerError("Cannot replace container at index {index} with a container that is not a DefinitionContainer".format(index = index))
+        elif container != self._empty_instance_container and container.getMetaDataEntry("type") != expected_type:
+            raise Exceptions.InvalidContainerError("Cannot replace container at index {index} with a container that is not of {type} type, but {actual_type} type.".format(index = index, type = expected_type, actual_type = container.getMetaDataEntry("type")))
+
+        super().replaceContainer(index, container, postpone_emit)
+
+    ##  Overridden from ContainerStack
+    #
+    #   This deserialize will make sure the internal list of containers matches with what we expect.
+    #   It will first check to see if the container at a certain index already matches with what we
+    #   expect. If it does not, it will search for a matching container with the correct type. Should
+    #   no container with the correct type be found, it will use the empty container.
+    #
+    #   \throws InvalidContainerStackError Raised when no definition can be found for the stack.
+    @override(ContainerStack)
+    def deserialize(self, contents: str) -> None:
+        super().deserialize(contents)
+
+        new_containers = self._containers.copy()
+        while(len(new_containers) < len(_ContainerIndexes.IndexTypeMap)):
+            new_containers.append(self._empty_instance_container)
```

Comment:
The extruder stack doesn't need to list all containers in the file. The rest are filled with empty containers. Some test cases actually test this behaviour.